events/service/src/main/java/org/projectnessie/events/service/EventService.java

```diff
@@ -0,0 +1,303 @@
+    LOGGER.debug("Received branch created result: {}", result);
+    fireEvent(factory.newReferenceCreatedEvent(result, repositoryId, user));
+  }
+
+  private void onReferenceAssignedResult(
+      ReferenceAssignedResult result, String repositoryId, @Nullable Principal user) {
+    LOGGER.debug("Received reference assigned result: {}", result);
+    fireEvent(factory.newReferenceUpdatedEvent(result, repositoryId, user));
+  }
+
+  private void onReferenceDeletedResult(
+      ReferenceDeletedResult result, String repositoryId, @Nullable Principal user) {
+    LOGGER.debug("Received reference deleted result: {}", result);
+    fireEvent(factory.newReferenceDeletedEvent(result, repositoryId, user));
+  }
+
+  private void fireCommitEvent(
+      Commit commit, BranchName targetBranch, String repositoryId, @Nullable Principal user) {
+    fireEvent(factory.newCommitEvent(commit, targetBranch, repositoryId, user));
+    if (hasContentSubscribers()) {
+      fireContentEvents(commit, targetBranch, repositoryId, user);
+    }
+  }
+
+  private void fireMergeEvent(
+      MergeResult<Commit> result, String repositoryId, @Nullable Principal user) {
+    fireEvent(factory.newMergeEvent(result, repositoryId, user));
+    if (hasCommitSubscribers()) {
+      for (Commit commit : result.getCreatedCommits()) {
+        fireCommitEvent(commit, result.getTargetBranch(), repositoryId, user);
+      }
+    }
+  }
+
+  private void fireTransplantEvent(
+      MergeResult<Commit> result, String repositoryId, @Nullable Principal user) {
+    fireEvent(factory.newTransplantEvent(result, repositoryId, user));
+    if (hasCommitSubscribers()) {
+      for (Commit commit : result.getCreatedCommits()) {
+        fireCommitEvent(commit, result.getTargetBranch(), repositoryId, user);
+      }
+    }
+  }
+
+  private void fireContentEvents(
+      Commit commit, BranchName targetBranch, String repositoryId, @Nullable Principal user) {
+    List<Operation> operations = commit.getOperations();
+    if (operations != null && !operations.isEmpty()) {
+      Hash hash = Objects.requireNonNull(commit.getHash());
+      for (org.projectnessie.versioned.Operation operation : operations) {
+        if (operation instanceof org.projectnessie.versioned.Put) {
+          ContentKey contentKey = ContentMapping.map(operation.getKey());
+          Content content = ContentMapping.map(((Put) operation).getValue());
+          fireEvent(
+              factory.newContentStoredEvent(
+                  targetBranch, hash, contentKey, content, repositoryId, user));
+        } else if (operation instanceof org.projectnessie.versioned.Delete) {
+          ContentKey contentKey = ContentMapping.map(operation.getKey());
+          fireEvent(
+              factory.newContentRemovedEvent(targetBranch, hash, contentKey, repositoryId, user));
+        }
+      }
+    }
+  }
+
+  private boolean hasContentSubscribers() {
+    return subscribers.hasSubscribersFor(EventType.CONTENT_STORED)
```

Comment:
Guess this can one and the function below can be memoized in `start()`.
Or at least have it end in `EnumSet.containsAll` which ends in a single bit-mask-check.